common/lib/core-interfaces/src/provider.ts

```diff
@@ -0,0 +1,70 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+/**
+ * @internal
+ * This utility type is meant for internal use by @see FluidObject
+ * Produces a valid FluidObject key given a type and a property.
+ * A valid FluidObject key is a property that exists on the incoming type
+ * as well as on the type of the property itself. For example, IProvideFoo.IFoo.IFoo
+ * This aligns with the FluidObject pattern expected to be used with all FluidObjects.
+ * For example:
+ * ```
+ * interface IProvideFoo{
+ *  IFoo: IFoo
+ * }
+ * interface IFoo extends IProvideFoo{
+ *  foobar();
+ * }
+ * ```
+ * This pattern enables discovery, and delegation in a standard way which is central
+ * to FluidObject pattern
+ */
+ export type FluidObjectProviderKeys<T, TProp extends keyof T = keyof T> =
+    string extends TProp ? never : number extends TProp? never : // exclude indexers [key:string |number]: any
+    TProp extends keyof Exclude<T[TProp], undefined> // TProp is a property of T, and T[TProp]
+        ? TProp
+        :never;
+
+/**
+ * This utility type take interface(s) that follow the FluidObject pattern, and produces
+ * a new type that can be used for inspection and discovery of those interfaces.
+ *
+ * It is meant to be used with types that are known to implement the FluidObject pattern.
+ * A common way to specify a type implements the FluidObject pattern is to expose it as a
+ * Provider without a generic argument.
+ *
+ * For example, if we have an interface like below
+ * ```
+ * interface IProvideFoo{
+ *  IFoo: IFoo
+ * }
+ * interface IFoo extends IProvideFoo{
+ *  foobar();
+ * }
+ * ```
+ *
+ * and a function that returns a FluidObject. You would do the following
+ *
+ * `const maybeFoo : Provider<IFoo> = getFluidObject()`;
```

Comment:
```suggestion
 * `const maybeFoo: FluidObject<IFoo> = getFluidObject()`;
```

Also -- I was thinking about suggested naming when using this stuff...  I think `maybeFoo` works if you don't think too hard about it, but will lead to a "wait, what?" moment when you look twice at `maybeFoo?.IFoo.doStuff` and wonder why it's not `maybeFoo?.doStuff`.

So additional suggestion would be to update this to `const mayProvideFoo: FluidObject<IFoo>` or even simply `const fluidObject: IFluidObject<IFoo>` (that works better if you do a union of interfaces).